tests/conftest.py

```diff
@@ -25,26 +50,391 @@ def config(kube_ns):
+        return "{command} exited with status {exit_code}: {message}".format(
+            command=self.command,
+            exit_code=self.exit_code,
+            message=self.message,
+        )
+
+
+def _exec_python_in_pod(kube_client, kube_ns, pod_name, code, kwargs=None, _retries=0):
+    """Run simple Python code in a pod
+
+    code can be a str of code, or a 'simple' Python function,
+    where source can be extracted (i.e. self-contained imports, etc.)
+
+    kwargs are passed to the function, if it is given.
+    """
+    if V(kubernetes.__version__) < V("11"):
+        pytest.skip(
+            f"exec tests require kubernetes >= 11, got {kubernetes.__version__}"
+        )
+    pod = wait_for_pod(kube_client, kube_ns, pod_name)
+    original_code = code
+    if not isinstance(code, str):
+        # allow simple self-contained (no globals or args) functions
+        func = code
+        code = "\n".join(
+            [
+                inspect.getsource(func),
+                "_kw = %r" % (kwargs or {}),
+                "{}(**_kw)".format(func.__name__),
+                "",
+            ]
+        )
+    elif kwargs:
+        raise ValueError("kwargs can only be passed to functions, not code strings.")
+
+    exec_command = [
+        "python3",
+        "-c",
+        code,
+    ]
+    print("Running {} in {}".format(code, pod_name))
+    # need to create ws client to get returncode,
+    # see https://github.com/kubernetes-client/python/issues/812
+    client = stream(
+        kube_client.connect_get_namespaced_pod_exec,
+        pod_name,
+        namespace=kube_ns,
+        command=exec_command,
+        stderr=True,
+        stdin=False,
+        stdout=True,
+        tty=False,
+        _preload_content=False,
+    )
+    client.run_forever(timeout=60)
+
+    # let pytest capture stderr
+    stderr = client.read_stderr()
+    print(stderr, file=sys.stderr)
+
+    returncode = client.returncode
+    if returncode:
+        print(client.read_stdout())
+        if _retries == 0:
+            raise ExecError(exit_code=returncode, message=stderr, command=code)
+        else:
+            # retry
+            time.sleep(1)
+            return _exec_python_in_pod(
+                kube_client,
+                kube_ns,
+                pod_name,
+                code,
+                _retries=_retries - 1,
+            )
+    else:
+        return client.read_stdout().rstrip()
+
+
+@pytest.fixture
+def exec_python_pod(kube_client, kube_ns):
+    """Fixture to return callable to execute python in a pod by name
+
+    Used as a fixture to contain references to client, namespace
+    """
+    return partial(_exec_python_in_pod, kube_client, kube_ns)
+
+
+@pytest.fixture(scope="session")
+def exec_python(kube_ns, kube_client):
+    """Return a callable to execute Python code in a pod in the test namespace
+
+    This fixture creates a dedicated pod for executing commands
+    """
+    pod_name = "kubespawner-test-exec"
```

Comment:
This fixture exists because we needed to talk to exactly one other pod, but that seems to no longer be the case as there is now always a hub pod to exec into, if we want.